csharp/LogicalRead.cs

```diff
@@ -292,11 +292,11 @@ public static class LogicalRead
             source.CopyTo(destination);
         }
 
-        public static void ConvertNative<TValue>(ReadOnlySpan<TValue> source, ReadOnlySpan<short> defLevels, Span<TValue?> destination, short nullLevel) where TValue : unmanaged
+        public static void ConvertNative<TValue>(ReadOnlySpan<TValue> source, ReadOnlySpan<short> defLevels, Span<TValue?> destination, short definedLevel) where TValue : unmanaged
         {
             for (int i = 0, src = 0; i < destination.Length; ++i)
             {
-                destination[i] = defLevels[i] == nullLevel ? default(TValue?) : source[src++];
+                destination[i] = defLevels.IsEmpty || defLevels[i] != definedLevel ? default(TValue?) : source[src++];
```

Comment:
> So the column data is still wrapped in a struct but one that only has a single msg field. Looking at the C++ parquet::arrow API there doesn't seem to be an equivalent to this, it only exposes one column which is the "nested" column so you must read all of the nested columns if you want any one of them.
> 
> So one way forward if we need to disambiguate between null structs and null inner values might be to provide a similar higher level API for reading structs that operates on a whole top level schema node rather than working at the leaf level.

Yep, I agree, I had discussions with Philip about this (off-Github) and the best move forward is some sort of `Nested<T>` type if we want to continue with the per-column reading approach. Or alternatively reading structs directly, as you stated, where you can check if the struct itself is nullable or inner members.

>Although for non-nullable native types we use the convert method override that doesn't take a defLevels argument. And for strings I don't think we allow non-nullable values, so in practice this might not be causing an issue.
>
> I'd argue that for cases where we have an override that doesn't take the defLevels span, we probably shouldn't be checking IsEmpty and can assume defLevels is always non-empty, as we did before, unless I'm missing some reason why the empty check is now required? But for the ConvertString and ConvertByteArray methods this logic should be fixed in case we start supporting required strings.

Yep, I hadn't accounted for null def levels, as this case is handled by the native reader. I'd probably say we should add the check to all types, to protect against any future implementation / disabling of overrides that might expose this bug?